### PR TITLE
Feat/geolocation/integrate gps alert screens

### PR DESCRIPTION
**Pull request by lazarinibruno**
# Integrate the GPS in the Alert Screens

## Description
To link the alerts to a location, both the `CreateAlert` and `EditAlert` have a location field. In this field, the user can choose between a location they enter or use their current location. While the logic for entering a location was already implemented by @taghizadlaura, there was no logic to get the GPS location.

Therefore, this PR introduces the logic necessary to fetch the device's GPS location using the `GPSService`  whenever the user selects the "Current location".

## Changes
The most important change is just a few lines of code in the `LocationField` composable in the file `AlertComponents.kt`. Whenever the user clicks on the "Current Location" drop-down menu item, the `location` StateFlow of `GPSServiceImpl` is fetched. Moreover, I changed the visibility of the drop-down menu to show the "Current Location" item as soon as the user clicks on the field (before, it only appeared when the user started typing).

There are many files changed, but its basically changing class signatures to take as argument the `GPSServiceImpl`.

Also, I updated the `C.kt` file to correctly distinguish between the test tags for:
- CreateAlert screen
- EditAlertScreen
- Shared tags

## Note
While writing the code, I noticed that it is redundant to have a `GPSLocation` and `Location` data class (the only difference being that the latter has a string for the location's name). I'm thinking of only using `Location` for the entirety of the code. However, as it would change code all over the place, I figured that it is better to have a dedicated PR (I already added the [draft](https://github.com/orgs/PeriodPals/projects/4/views/1?pane=issue&itemId=88523851) to the Scrum Board).

### Added
None.

### Modified
- app/src/main/java/com/android/periodpals/MainActivity.kt
- app/src/main/java/com/android/periodpals/model/location/GPSLocation.kt
- app/src/main/java/com/android/periodpals/resources/C.kt
- app/src/main/java/com/android/periodpals/ui/alert/CreateAlert.kt
- app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt
- app/src/main/java/com/android/periodpals/ui/components/AlertComponents.kt
- app/src/test/java/com/android/periodpals/ui/alert/CreateAlertScreenTest.kt
- app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt

### Removed
None.

## Dependencies Added
None.

## Testing

- I adapted the tests to take into account these changes. Namely, the tests verify that there are 4 drop-down items when there are suggestions (3 for the suggestions, 1 for the "Current Location"), and 1 drop-down item when there are no suggestions (for "Current location").
- I also added one more test that creates a valid alert by selecting the "Current Location" drop-down item.

### CODE REVIEW

**Review by francelu — APPROVED** [review]
## **Summary**
This PR integrates GPS functionality into the alert screens, allowing users to easily select their current location when creating or editing alerts. The change improves the UI by immediately offering the current location in the drop-down menu, which enhances user experience. Good job !


## **Important**

### **Code Quality**
- **Readability**: The changes are clear and easy to follow. The logic for fetching the GPS location is implemented concisely, and the update to the drop-down menu visibility is a good improvement.
- **Consistency**: The adaptation to take `GPSServiceImpl` as a parameter in relevant classes is consistent with the rest of the codebase. The change does not break any existing patterns, and it fits well within the existing architecture.

### **Functionality**
- **Correctness**: The logic for fetching and displaying the current location is solid. The change ensures that users can easily select their current location right away, which improves the UX. The tests validate that the new behaviour works as expected.
- **Edge Cases**: There are no visible edge cases at the moment. See code comment in `EditAlertScreenTest.kt` too.

### **Performance**
- No performance issues noted. The code remains efficient with minimal added complexity.

### **Security**
- No security concerns related to this PR.


## **Testing**
- See code comments (minor issue).
- Else, the tests have been adapted well to cover the new behaviour. The test cases validate the drop-down items correctly, and an additional test to ensure the "Current Location" item functions as expected is a great addition.


## **Documentation**
- See code comments on where it's still missing a little.
- Else, the PR description is clear and provides a good overview of the changes made. There is a good explanation of the modified logic and reasoning behind showing the "Current Location" item in the drop down menu right away.


## **Suggestions**
- Good work on adapting the tests for these changes.
- As mentioned in the PR description, consider consolidating `GPSLocation` and `Location` in a future PR if you find it adds clarity and simplifies the codebase.


## **Steps before PR merged**
Please check my code comments for minor clarifications. Once those are addressed, the PR will be good to go.

Great job! This update is a solid improvement to the alert creation and editing process.

File: app/src/main/java/com/android/periodpals/model/location/GPSLocation.kt
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/model/location/GPSLocation.kt:1]: Exactly!

File: app/src/main/java/com/android/periodpals/ui/alert/CreateAlert.kt [hunk 1]
```diff
@@ -60,6 +61,7 @@ private const val SUBMISSION_BUTTON_TEXT = "Ask for Help"
 fun CreateAlertScreen(
     navigationActions: NavigationActions,
     locationViewModel: LocationViewModel = viewModel(factory = LocationViewModel.Factory),
+    gpsService: GPSServiceImpl
```
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/alert/CreateAlert.kt:64]: Good catch!

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt [hunk 2]
```diff
@@ -78,6 +79,7 @@ fun EditAlertScreen(
             message = "Hello!",
             status = Status.CREATED,
             createdAt = ""), // TODO: remove this mock alert, for now it is used to visualize UI
+    gpsService: GPSServiceImpl
```
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt]: I put the `Alert` as the first parameter to have all the classes together. What do you think?

File: app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt:1]: Completely forgot about that! Thanks for bringing it up.

File: app/src/main/java/com/android/periodpals/ui/alert/CreateAlert.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/alert/CreateAlert.kt]: Good ! You did the necessary :)

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt]: LGTM. It's more logical as that's the most important parameter for this screen.  
However, the TODO needs to be moved too, it's for the Alert that is mocked :)

File: app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt:1]: - Why did you not start the test with the following as you did in the similar one for create alert ?  
  ```kotlin
    `when`(locationViewModel.query).thenReturn(MutableStateFlow(LOCATION_SUGGESTION1.name))
    `when`(locationViewModel.locationSuggestions).thenReturn(MutableStateFlow(emptyList()))
  ```
- I also noticed that most of the `testTag` are sth like `Tag.CreateAlertScreen.PRODUCT_FIELD`.
  I think it would be better if we had as there's for the AlertLists tags, an `object AlertInputField` in which we can put shared tags between create and edit alert screens

[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt:1]: > Why did you not start the test with the following as you did in the similar one for create alert ?

Since that is used for the location suggestions, which get only triggered when the user starts typing:

```kotlin
// In AlertComponents.kt, LocationField composable
onValueChange = {
          name = it
          locationViewModel.setQuery(it)
          showDropdown = true // Show dropdown when user starts typing
}
```

the mock is not necessary.

> I also noticed that most of the testTag are sth like Tag.CreateAlertScreen.PRODUCT_FIELD.
I think it would be better if we had as there's for the AlertLists tags, an object AlertInputField in which we can put shared tags between create and edit alert screens

Good idea :) I can do it.

File: app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/alert/EditAlert.kt]: All good :)

File: app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt:1]: > Good idea :) I can do it. I think it is enough to refactor `Tag.CreateAlertScreen` to `Tag.AlertScreens`, no? Since they are shared bu both `EditAlert` and `CreateAlert`.

I don't think so. You would still need to extract the screen and submit button out, since they aren't the same / similar on both screens.  
I would :
- Refactor the current `AlertScreens` to `AlertInputs` (since we also have an `AlertListsScreen`).
- Take out the `SCREEN` and `SUBMIT_BUTTON` tags into `CreateAlertScreen` object.
- Move the `EditAlertScreen`before the `AlertListsScreen` because it's linked to `AlertInputs` and `CreateAlertScreens`.
But do as you feel is the best :)

File: app/src/main/java/com/android/periodpals/resources/C.kt [hunk 3]
```diff
@@ -10,15 +10,16 @@ object C {
...
+      const val SCREEN = "createAlertScreen"
       const val INSTRUCTION_TEXT = "instructionText"
       const val PRODUCT_FIELD = "productField"
       const val URGENCY_FIELD = "urgencyField"
       const val DROPDOWN_ITEM = "dropdownItem"
       const val LOCATION_FIELD = "locationField"
       const val MESSAGE_FIELD = "messageField"
       const val SUBMIT_BUTTON = "submitButton"
+      const val CURRENT_LOCATION = "currentLocation"
     }
```
Comment (taghizadlaura) [inline app/src/main/java/com/android/periodpals/resources/C.kt:23]: I don't quite understand the name of the object: if it's meant for all alert screens, why doesn't it include the tags for the `EditAlertScreen` as well? And how would you differentiate the `SCREEN` tags? Moreover, you didn't update the related documentation. Please clarify this part.

**Review by taghizadlaura — APPROVED** [review]
## Summary
LGTM :
The PR introduces GPS functionality to the `CreateAlert` and `EditAlert` screens, allowing users to select their current location directly. This addition enhances the user experience by integrating GPS location retrieval into the alert creation and editing workflow.

## Important

### Code Quality
LGTM :
- The code is readable and consistent, particularly in how the `LocationField` composable is updated to handle the "Current Location" option.

### Functionality
LGTM :
- The implementation appears correct, addressing the core requirement of fetching the GPS location when selected.

### Performance
- No performance issues.

### Security
- No security issues.

## Testing
LGTM :
- Tests have been adapted to verify the new functionality, including the presence of drop-down items and selecting the "Current Location" option.

## Documentation
- Unless the documentation for the `AlertScreens` object for tags is not updated, the rest is correctly documented.

## Suggestions
- Consider clarifying the name of the object: if it is meant for all alert screens, why doesn't it include the tags for the `EditAlertScreen` as well? How would you differentiate the `SCREEN` tags?  
- Update the related documentation for the `AlertScreens` object to reflect these points.

## Steps Before PR Approval
LGTM but please answer my comments !

File: app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt
[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt:1]: Okay, you are right. I did not pay close attention to it. I updated the Tags to properly distinguish between CreateAlert, EditAlert and shared tags.

[the remark this answers is not in the document]
Comment (francelu) [reply app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt:1]: It's perfect !! :) Good job overall !!!

[the remark this answers is not in the document]
Comment (lazarinibruno) [reply app/src/test/java/com/android/periodpals/ui/alert/EditAlertScreenTest.kt:1]: Thank you :D
